Content.Server/GameObjects/EntitySystems/Atmos/GasTileOverlaySystem.cs

```diff
@@ -0,0 +1,350 @@
+        public void Invalidate(GridId gridIndex, MapIndices indices)
+        {
+            if (!_invalidTiles.TryGetValue(gridIndex, out var existing))
+            {
+                existing = new HashSet<MapIndices>();
+                _invalidTiles[gridIndex] = existing;
+            }
+
+            existing.Add(indices);
+        }
+
+        private GasOverlayChunk GetOrCreateChunk(GridId gridIndex, MapIndices indices)
+        {
+            if (!_overlay.TryGetValue(gridIndex, out var chunks))
+            {
+                chunks = new Dictionary<MapIndices, GasOverlayChunk>();
+                _overlay[gridIndex] = chunks;
+            }
+
+            var chunkIndices = GetGasChunkIndices(indices);
+
+            if (!chunks.TryGetValue(chunkIndices, out var chunk))
+            {
+                chunk = new GasOverlayChunk(gridIndex, chunkIndices);
+                chunks[chunkIndices] = chunk;
+            }
+
+            return chunk;
+        }
+
+        private void OnGridRemoved(GridId gridId)
+        {
+            if (_overlay.ContainsKey(gridId))
+            {
+                _overlay.Remove(gridId);
+            }
+        }
+
+        public void ResettingCleanup()
+        {
+            _invalidTiles.Clear();
+            _overlay.Clear();
+
+            foreach (var (_, data) in _knownPlayerChunks)
+            {
+                data.Reset();
+            }
+        }
+
+        private void OnPlayerStatusChanged(object sender, SessionStatusEventArgs e)
+        {
+            if (e.NewStatus != SessionStatus.InGame)
+            {
+                if (_knownPlayerChunks.ContainsKey(e.Session))
+                {
+                    _knownPlayerChunks.Remove(e.Session);
+                }
+
+                return;
+            }
+
+            if (!_knownPlayerChunks.ContainsKey(e.Session))
+            {
+                _knownPlayerChunks[e.Session] = new PlayerGasOverlay();
+            }
+        }
+
+        /// <summary>
+        ///     Checks whether the overlay-relevant data for a gas tile has been updated.
+        /// </summary>
+        /// <param name="gam"></param>
+        /// <param name="oldTile"></param>
+        /// <param name="indices"></param>
+        /// <param name="overlayData"></param>
+        /// <returns>true if updated</returns>
+        private bool TryRefreshTile(GridAtmosphereComponent gam, GasOverlayData oldTile, MapIndices indices, out GasOverlayData overlayData)
+        {
+            var tile = gam.GetTile(indices);
+            var tileData = new List<GasData>();
+
+            for (var i = 0; i < Atmospherics.TotalNumberOfGases; i++)
+            {
+                var gas = Atmospherics.GetGas(i);
+                var overlay = Atmospherics.GetOverlay(i);
+                if (overlay == null || tile.Air == null) continue;
+
+                var moles = tile.Air.Gases[i];
+
+                if (moles == 0.0f || moles < gas.GasMolesVisible) continue;
```

Comment:
```suggestion
                if (moles < gas.GasMolesVisible) continue;
```

Just remove this comparison, it won't do much.